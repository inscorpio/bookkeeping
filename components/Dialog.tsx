import type { ReactNode } from 'react'
import { useEffect, useRef } from 'react'

interface Props {
  visible?: boolean
  confirm?: ReactNode
  children?: ReactNode
  onClose: () => void
}

export default function Dialog(
  {
    visible = false,
    confirm,
    children,
    onClose,
  }: Props,
) {
  const dialogRef = useRef<HTMLDialogElement>(null)
  useEffect(() => {
    if (dialogRef.current) {
      // I don't known why it throws an InvalidStateError error on a mobile phone without try...catch...
      try {
        visible
          ? dialogRef.current.showModal()
          : dialogRef.current.close()
      }
      catch {}
    }
  }, [visible])

  return (
    <dialog
      ref={dialogRef}
      className="w-screen max-w-none m-0 mt-auto text-stone-950 outline-none"
    >
      <div className="flex flex-col w-full h-[248px] py-3">
        <div className="flex justify-between items-center px-3">
          <button onClick={onClose}>取消</button>
          {confirm}
        </div>
        <div className="flex-1 mt-2 overflow-hidden">
          {children}
        </div>
      </div>
    </dialog>
  )
}
