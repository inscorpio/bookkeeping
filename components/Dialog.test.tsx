import { render, screen } from '@testing-library/react'
import { expect, test } from 'vitest'
import Dialog from '~/components/Dialog'

test('initial', async () => {
  render(
    <Dialog
      visible={false}
      onClose={() => {}}
    />,
  )
  const dialog = screen.getByRole('dialog')
  expect(dialog).not.toBeVisible()
})
