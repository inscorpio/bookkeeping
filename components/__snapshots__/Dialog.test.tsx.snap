// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render 1`] = `
<div>
  <dialog
    class="w-screen max-w-none m-0 mt-auto text-stone-950 outline-none"
  >
    <div
      class="flex flex-col w-full h-[248px] py-3"
    >
      <div
        class="flex justify-between items-center px-3"
      >
        <button>
          取消
        </button>
      </div>
      <div
        class="flex-1 mt-2 overflow-hidden"
      />
    </div>
  </dialog>
</div>
`;
